### PR TITLE
Set Up

### PR DESCRIPTION
**Pull request by geetikabatra**
Signed-off-by: Geetika Batra <gbatra@redhat.com>

### CODE REVIEW

File: run-linter.sh [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+if [ $fail -eq 0 ]
+then
+    echo "All checks passed for $pass source files"
+else
+    let total=$pass+$fail
+    echo "Linter fail, $fail source files out of $total source files need to be fixed"
+    exit 1
+fi
+© 2018 GitHub, Inc.
```
Comment (sara-02) [inline run-linter.sh]: Un-intended copy paste ;)

**Review by sara-02 — CHANGES_REQUESTED** [review]

File: run-linter.sh [hunk 2]
```diff
@@ -0,0 +1,105 @@
+Skip to content
```
Comment (sara-02) [inline run-linter.sh]: I think this section as well.

**Review by sara-02 — CHANGES_REQUESTED** [review]

File: run-linter.sh (hunk 1 above)
Comment (geetikabatra) [reply run-linter.sh]: Thanks @sara-02, fixing it :)

File: f8a_version_comparator/base.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+
+from abc import ABCMeta, abstractmethod
+
+
+class Base(metaclass=ABCMeta):
+    """Base class for maven version comparator tasks."""
+
+    @abstractmethod
+    def compare_to(self):
```
Comment (geetikabatra) [inline f8a_version_comparator/base.py]: @msrb and @tisnik. pydocstyle fails here with error `Missing docstring in public method` . But if I add a docstring, it throws `Cannot parse file`. Any idea why?

Comment (msrb) [reply f8a_version_comparator/base.py]: @GeetikaBatra how does the docstring look like?

File: f8a_version_comparator/base.py [hunk 4]
```diff
@@ -26,9 +26,9 @@ class Base(metaclass=ABCMeta):
 
     @abstractmethod
     def compare_to(self):
+    	"""Compare two maven versions."""
```
Comment (geetikabatra) [inline f8a_version_comparator/base.py]: @msrb, Like this.

File: f8a_version_comparator/base.py (hunk 3 above)
Comment (msrb) [reply f8a_version_comparator/base.py]: I think you're mixing spaces and tabs on that docstring line.

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    name='f8a_version_comparator',
+    version='0.1',
+    packages=find_packages(exclude=['tests', 'tests.*']),
+    install_requires=reqs,
+    dependency_links=dep_links,
+    include_package_data=True,
+    author='Geetika Batra',
+    author_email='gbatra@redhat.com',
+    description='fabric8-analytics maven version comparator',
+    license='GPLv3',
```
Comment (CermakM) [inline setup.py]: I am not sure this is a good idea to use such prohibitive license. A lot of our projects use Apache 2.0 (and so does some of which can make use of this version comparator) which is only one-way compatible with GPLv3. Surely, those are our projects, but still.. Might be worth consideration.

Comment (msrb) [reply setup.py]: It's generally a bad idea to license a library under GPL. LGPL would be a much better option.

But like you said, we use ASL 2.0 in most of our projects. So I'd recommend using it here as well.

Comment (geetikabatra) [reply setup.py]: I agree @msrb and @CermakM, will change it :)

**Comment by geetikabatra** [comment]
@msrb and @tisnik PTAL. All the new classes introduced are equivalent to java implementation of object classes in maven versioning.
Following was the first attempt to implement it using stacks, but I couldn't reach to a proper solution, so instead I started implementing it the way it is implemented in java.
https://github.com/fabric8-analytics/fabric8-analytics-version-comparator/pull/1/files#diff-e99bc216b10f147527267638624f749c

File: cico_build_deploy.sh [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+set -ex
+
+. cico_setup.sh
+
+build_image
+
+./runtests.sh
+
+push_image
+
```
Comment (msrb) [inline cico_build_deploy.sh]: There will be no docker image for this library so we can remove `push_image` and `build_image`.

File: cico_run_tests.sh [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+set -ex
+
+. cico_setup.sh
+
+# not needed for tests, but we can check that the image actually builds
+build_image
+
+./runtests.sh
+
+push_image
```
Comment (msrb) [inline cico_run_tests.sh]: `build_image` and `push_image` are not needed.

File: cico_setup.sh [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    fi
+}
+
+prep() {
+    yum -y update
+    yum -y install docker git which epel-release python34-virtualenv postgresql
+    yum -y install python34-pip
+    pip3 install virtualenv
+    pip3 install docker-compose
+    systemctl start docker
```
Comment (msrb) [inline cico_setup.sh]: I think we don't need docker, docker-compose, postgresql.

File: cico_setup.sh [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+build_image() {
```
Comment (msrb) [inline cico_setup.sh]: Not needed.

File: cico_setup.sh [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    make docker-build
+}
+
+tag_push() {
```
Comment (msrb) [inline cico_setup.sh]: Not needed.

File: cico_setup.sh [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    local target=$1
+    local source=$2
+    docker tag ${source} ${target}
+    docker push ${target}
+}
+
+push_image() {
```
Comment (msrb) [inline cico_setup.sh]: Not needed.

File: f8a_version_comparator/base.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env python
```
Comment (msrb) [inline f8a_version_comparator/base.py]: shebang not needed as this is not a standalone script.

File: f8a_version_comparator/base.py [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+# Copyright © 2018 Red Hat Inc.
+#
+# This program is free software: you can redistribute it and/or modify
```
Comment (msrb) [inline f8a_version_comparator/base.py]: This is a license header of GPLv3+, but `setup.py` declares license as `ASL 2.0`. So which one is correct? :)

File: f8a_version_comparator/base.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+from abc import ABCMeta, abstractmethod
+
+
+class Item(metaclass=ABCMeta):
+    """Base class for maven version comparator tasks."""
+
+    @abstractmethod
+    def compare_to(self, item):
+        """Compare two maven versions."""
+        raise NotImplementedError()
```
Comment (msrb) [inline f8a_version_comparator/base.py]: I think you can say just `pass` here as it shouldn't be possible to instantiate class with abstract methods anyway.

File: f8a_version_comparator/base.py [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+    @abstractmethod
+    def is_none(self):
+        """Check if none."""
+        raise NotImplementedError()
```
Comment (msrb) [inline f8a_version_comparator/base.py]: Same here, `pass` should do the trick.

File: runtests.sh [hunk 16]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash -ex
```
Comment (msrb) [inline runtests.sh:1]: The file is missing executable bit.

`chmod +x runtests.sh`

File: run-linter.sh [hunk 17]
```diff
@@ -0,0 +1,73 @@
+directories="f8a_version_comparator tests"
```
Comment (msrb) [inline run-linter.sh:2]: The file is missing executable bit.

`chmod +x run-linter.sh`

File: check-docstyle.sh [hunk 18]
```diff
@@ -0,0 +1,72 @@
+directories="f8a_version_comparator tests"
```
Comment (msrb) [inline check-docstyle.sh:2]: The file is missing executable bit.

`chmod +x check-docstyle.sh` should do the trick.

**Review by msrb — CHANGES_REQUESTED** [review]
First quick review, I will have more comments later :wink: Please take a look.

File: f8a_version_comparator/base.py (hunk 14 above)
Comment (tisnik) [reply f8a_version_comparator/base.py]: Actually I'd suggest to make the function empty, ie. put the docstring in it, nothing more :)
Instead you won't be able to cover it by unit tests.

File: f8a_version_comparator/base.py (hunk 15 above)
Comment (tisnik) [reply f8a_version_comparator/base.py]: (or nothing :)

File: f8a_version_comparator/base.py (hunk 14 above)
Comment (geetikabatra) [reply f8a_version_comparator/base.py]: Thanks @msrb and @tisnik for the reviews :). One quick question. These are abstract method definitions, If I just put a docstring and if this method is later on not implemented, it won't raise any error. Please correct me if I am wrong. :)

Comment (CermakM) [reply f8a_version_comparator/base.py]: @msrb @GeetikaBatra it does, see the [gist](https://gist.github.com/CermakM/5cc4b6fd5da3b71135ce09fdee17d9e6) ... the only problem is it raises TypeError .. Personaly, I usually do write `raise NotImplementedError()`, since it is more explicit.

Comment (geetikabatra) [reply f8a_version_comparator/base.py]: Thanks @CermakM. Looks like changing it to pass will also work fine. :)

**Comment by geetikabatra** [comment]
@msrb, This PR is complete and more than 50 cases are tested, all of them work perfectly fine.  https://github.com/fabric8-analytics/fabric8-analytics-version-comparator/pull/1/files#diff-eaf7c8089806c6cdbc2c95d1f41238ea

This PR can be merged now :)

File: runtests.sh (hunk 16 above)
Comment (geetikabatra) [reply runtests.sh:1]: @msrb Do we need this, I didn't see this command in worker. :)

Comment (msrb) [reply runtests.sh:1]: It's there in worker ;)

```bash
$ ls -la runtest.sh
-rwxrwxr-x. 1 michal michal 3897 Apr 11 20:51 runtest.sh
```

But it's missing here:
```bash
$ ls -la runtests.sh 
-rw-rw-r--. 1 michal michal 694 May  6 17:30 runtests.sh
```

Running `chmod +x runtests.sh` should do the trick here.

File: f8a_version_comparator/__init__.py [hunk 19]
```diff
@@ -0,0 +1,19 @@
+# Copyright © 2018 Red Hat Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
```
Comment (msrb) [inline f8a_version_comparator/__init__.py]: This is GPLv3+ license header, however `setup.py` declared license as ASL 2.0. So this header should be replaced with https://www.apache.org/licenses/LICENSE-2.0#apply.

File: setup.py [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+    name='f8a_version_comparator',
+    version='0.1',
+    packages=find_packages(exclude=['tests', 'tests.*']),
+    install_requires=reqs,
+    dependency_links=dep_links,
+    include_package_data=True,
+    author='Geetika Batra',
+    author_email='gbatra@redhat.com',
+    description='fabric8-analytics maven version comparator',
+    license=' ASL 2.0',
```
Comment (msrb) [inline setup.py]: Please remove the whitespace.

File: f8a_version_comparator/comparable_version.py [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+
+"""Module to implement Comparable Version class."""
+
+from .item_object import IntegerItem
+from .item_object import StringItem
+from .item_object import ListItem
+
+
+class ComparableVersion():
+    """Class for Comaparable Version."""
```
Comment (msrb) [inline f8a_version_comparator/comparable_version.py]: Typo: Comaparable - > Comparable

File: f8a_version_comparator/comparable_version.py [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+
+    def parse_item(self, _is_digit, buf):
+        """Wrap items in version in respective object class."""
+        if _is_digit:
+            return IntegerItem(buf)
+
+        return StringItem(buf, False)
+
+    def compare_to(self, obj):
+        """Compare two comaparable version objects."""
```
Comment (msrb) [inline f8a_version_comparator/comparable_version.py]: Typo in "comaparable" :)

File: f8a_version_comparator/item_object.py [hunk 23]
```diff
@@ -0,0 +1,219 @@
...
+# TODO: setup logging
+
+
+class IntegerItem(Item):
+    """Integer Item class for maven version comparator tasks."""
+
+    def __init__(self, str_version):
+        """Initialize integer from string value of version.
+
+        :str_value: part of version supplied as string
```
Comment (msrb) [inline f8a_version_comparator/item_object.py]: The param is called `str_version`.

File: f8a_version_comparator/item_object.py [hunk 24]
```diff
@@ -0,0 +1,219 @@
...
+        return False
+
+
+class StringItem(Item):
+    """String Item class for maven version comparator tasks."""
+
+    def __init__(self, str_version, followed_by_digit):
+        """Initialize string value of version.
+
+        :str_value: part of version supplied as string
```
Comment (msrb) [inline f8a_version_comparator/item_object.py:67]: params are `str_version` and `followed_by_digit`.

File: f8a_version_comparator/item_object.py [hunk 25]
```diff
@@ -0,0 +1,219 @@
...
+        if isinstance(item, ListItem):
+            return -1
+        else:
+            raise ValueError("invalid item" + type(item))
+
+    def to_string(self):
+        """Return value in string form."""
+        return str(self.value)
+
+    @classmethod
```
Comment (msrb) [inline f8a_version_comparator/item_object.py]: I think this method shouldn't be a class method.

File: f8a_version_comparator/item_object.py [hunk 26]
```diff
@@ -0,0 +1,219 @@
...
+    def is_none(self):
+        """Check if none."""
+        if self.value is release_version_index and self.value is 0:
```
Comment (msrb) [inline f8a_version_comparator/item_object.py]: What does this IF statement do exactly?

Comment (msrb) [reply f8a_version_comparator/item_object.py]: Note it seems to me that this method is never called, do we need it?

File: f8a_version_comparator/item_object.py [hunk 27]
```diff
@@ -0,0 +1,219 @@
...
+                if result is not 0:
+                    return result
+
+            return 0
+        else:
+            raise ValueError("invalid item" + type(item))
+
+    def is_none(self):
+        """Check if none."""
+        if len(self.arraylist) is o:
```
Comment (msrb) [inline f8a_version_comparator/item_object.py]: `o` is not declared anywhere?

We don't call this method anywhere, do we need it?

File: f8a_version_comparator/item_object.py [hunk 28]
```diff
@@ -0,0 +1,219 @@
...
+    def get_list(self):
+        """Get object list items."""
+        return self.array_list
+
+    def normalize(self):
+        """Remove trailing items: 0, "", empty list."""
+        red_list = [0, None, ""]
+        i = len(self.array_list) - 1
+        while(i >= 0):
+            lastItem = self.array_list[i]
```
Comment (msrb) [inline f8a_version_comparator/item_object.py]: We use snake_case everywhere in the code, so camelCase like `lastItem` shouldn't be here.

File: f8a_version_comparator/item_object.py [hunk 29]
```diff
@@ -0,0 +1,219 @@
...
+
+            if(not isinstance(lastItem, ListItem)):
```
Comment (msrb) [inline f8a_version_comparator/item_object.py]: Those parentheses are superfluous.

File: f8a_version_comparator/item_object.py [hunk 30]
```diff
@@ -0,0 +1,219 @@
...
+
+        if isinstance(item, IntegerItem):
+            return -1
+        if isinstance(item, StringItem):
+            return 1
+        if isinstance(item, ListItem):
+            left_iter = iter(self.array_list)
+            right_iter = iter(item.get_list())
+
+            while(True):
```
Comment (msrb) [inline f8a_version_comparator/item_object.py]: No need for the parentheses.

File: f8a_version_comparator/item_object.py [hunk 31]
```diff
@@ -0,0 +1,219 @@
...
+
+    def get_list(self):
+        """Get object list items."""
+        return self.array_list
+
+    def normalize(self):
+        """Remove trailing items: 0, "", empty list."""
+        red_list = [0, None, ""]
+        i = len(self.array_list) - 1
+        while(i >= 0):
```
Comment (msrb) [inline f8a_version_comparator/item_object.py]: No need for the parentheses.

**Review by msrb — CHANGES_REQUESTED** [review]
More comments, PTAL :wink:

File: tests/conftest.py [hunk 32]
```diff
@@ -0,0 +1,27 @@
+# Copyright © 2018 Red Hat Inc.
```
Comment (msrb) [inline tests/conftest.py]: Seems to me that we don't need this file at all (?)

File: f8a_version_comparator/item_object.py (hunk 27 above)
Comment (geetikabatra) [reply f8a_version_comparator/item_object.py]: We are not using this method, I can remove it :)

**Comment by geetikabatra** [comment]
@msrb, I have fixed the issues. PTAL :)

File: tests/test_version_comparator.py [hunk 33]
```diff
@@ -0,0 +1,106 @@
...
+    assert(res == 0)
+
+
+def test_comparisons():
+    """Test function covering all the cases."""
+    check_version_order("1-alpha-1", "1.0")
+    check_version_order("1", "2")
+    check_version_order("1.5", "2")
+    check_version_order("1", "2.5")
+    check_version_order("1", "2")
```
Comment (CermakM) [inline tests/test_version_comparator.py]: Same test on line :42, if I am not confused by all these checks :)

File: f8a_version_comparator/comparable_version.py [hunk 34]
```diff
@@ -0,0 +1,108 @@
...
+                    temp = ListItem()
+                    ref_list.add_item(temp)
+                    ref_list = temp
+                    self.parse_stack.append(ref_list)
+                _is_digit = False
+
+        if len(version) > _start_index:
+            ref_list.add_item(self.parse_item(_is_digit, version[_start_index:]))
+
+        while len(self.parse_stack) > 0:
```
Comment (CermakM) [inline f8a_version_comparator/comparable_version.py]: You can compare just like `while self.parse_stack:` .. in fact this is recommended by PEP8 standard.

File: f8a_version_comparator/item_object.py [hunk 35]
```diff
@@ -0,0 +1,199 @@
...
+    """String Item class for maven version comparator tasks."""
+
+    def __init__(self):
+        """Initialize string value of version."""
+        self.array_list = list()
+
+    def add_item(self, item):
+        """Add item to array list."""
+        arr_list = self.array_list
+        arr_list.append(item)
```
Comment (CermakM) [inline f8a_version_comparator/item_object.py]: why not `self.arr_list.append(item)`?

File: tests/test_version_comparator.py [hunk 36]
```diff
@@ -0,0 +1,106 @@
...
+
+from f8a_version_comparator.comparable_version import ComparableVersion
+
+
+def check_version_order(v1, v2):
+    """Check order of versions."""
+    c = ComparableVersion(v1)
+    c1 = ComparableVersion(v2)
+    res = c.compare_to(c1)
+    assert(res == -1)
```
Comment (CermakM) [inline tests/test_version_comparator.py]: Might be useful to add some message to the assert, like `f"Expected -1, got {res}"` to see the actual returned value on case of error.

File: tests/test_version_comparator.py [hunk 37]
```diff
@@ -0,0 +1,106 @@
...
+
+
+def check_version_equal(v1, v2):
+    """Check if versions are equal."""
+    c = ComparableVersion(v1)
+    c1 = ComparableVersion(v2)
+    res = c.compare_to(c1)
+    assert(res == 0)
```
Comment (CermakM) [inline tests/test_version_comparator.py]: Same as above :)

File: f8a_version_comparator/item_object.py [hunk 38]
```diff
@@ -0,0 +1,199 @@
...
+        else:
+            raise ValueError("invalid item" + type(item))
+
+    def to_string(self):
+        """Return value in string form."""
+        return str(self.value)
+
+
+class ListItem(Item):
+    """String Item class for maven version comparator tasks."""
```
Comment (CermakM) [inline f8a_version_comparator/item_object.py]: Copied docstring from the StringItem :)

File: f8a_version_comparator/item_object.py [hunk 39]
```diff
@@ -0,0 +1,199 @@
...
+    def __init__(self):
+        """Initialize string value of version."""
+        self.array_list = list()
+
+    def add_item(self, item):
+        """Add item to array list."""
+        arr_list = self.array_list
+        arr_list.append(item)
+
+    def get_list(self):
```
Comment (CermakM) [inline f8a_version_comparator/item_object.py]: This getter is not really necessary, unless you want to make `self.array_list` private to the class

File: f8a_version_comparator/comparable_version.py [hunk 40]
```diff
@@ -0,0 +1,108 @@
...
+
+class ComparableVersion():
+    """Class for Comparable Version."""
+
+    def __init__(self, version):
+        """Initialize comparable version class.
+
+        :version: Version supplied as a string
+        """
+        self.parse_version(version)
```
Comment (CermakM) [inline f8a_version_comparator/comparable_version.py]: It is a good practice to predefine the class attributes in the `__init__` function and just assign the desired values afterwards in the `parse_version`.

**Review by CermakM — CHANGES_REQUESTED** [review]
Additional comments:
- it would be useful to implement `__str__` representations of the objects .. for example, if I want to see the value of `ver` which I instantiated such as
```python
ver = ComparableVersion('1.0.0')
```
I have to do:
```python
[v.value for v in ver.items.array_list]
```
 and if I wanted the same for
```python
ver = ComparableVersion('1.0.0-a')
```
I'd have to figure out how to get around the `AttributeError`

- it would be useful if there were `compare` methods which accept strings, instead of having to instantiate the ComparableVersion

- `compare_to` method is very unintuitive, it would be great if there were for example `is_lower`, `is_greater`, `is_equal` functions.

**Comment by geetikabatra** [comment]
@CermakM, the logic for version comparison is a little complicated. First iteration involved passing string to compare function instead of Comparable version instantiation, but I was not able to cover the edge cases which invlolved '-' and aliases.
So I thought better of it and implemented it the way it is done in Java.
IMO, compare_to method should be fine :)

File: tests/test_version_comparator.py (hunk 33 above)
Comment (geetikabatra) [reply tests/test_version_comparator.py]: Fixed

File: f8a_version_comparator/item_object.py (hunk 39 above)
Comment (geetikabatra) [reply f8a_version_comparator/item_object.py]: @CermakM, We can merge this PR and, I can create another PR to fix the get_list method. WDYT.
@msrb  ^ :)

File: f8a_version_comparator/comparable_version.py (hunk 40 above)
Comment (geetikabatra) [reply f8a_version_comparator/comparable_version.py]: We would then have to call parse_version explicitly from class object.

File: f8a_version_comparator/item_object.py (hunk 39 above)
Comment (CermakM) [reply f8a_version_comparator/item_object.py]: @GeetikaBatra Sure, it can be done in the future :)

File: f8a_version_comparator/comparable_version.py (hunk 40 above)
Comment (CermakM) [reply f8a_version_comparator/comparable_version.py]: Oh well, I didn't mean to change the call of `parse_version`, just that it is good practice to declare the class attributes directly in `__init__` function, since it can break the attribute hints in some IDEs later... 

as an Example:
```python
class Foo:
  def __init__(self):
    self.foo_attr = None
    self.parse()
  def parse(self):
    self.foo_attr = "bar"
```

the `foo_attr` is declared directly in the `__init__` instead of in the `parse` method.

**Comment by geetikabatra** [comment]
@CermakM and @msrb. Please take a look :)

**Comment by msrb** [comment]
LGTM, thanks :wink:
